content/blog/posts/2022-08-04-add-desktop-notifications.md

```diff
@@ -0,0 +1,241 @@
+    });
+  }
+}
+```
+
+This one is a bit more straightforward. We check permission, as noted in the previous code example, and as long as we're not denied, a notification can be shown. There's really only one required argument, a title. Here we are adding a basic option for body text as well. This will just show a notification and the user can dismiss it or select to go to browser settings, if they'd like to change the permissions. Let's make sure our instance fires off when loaded before we move on.
+
+```js
+class NotificationService {
+  constructor() {
+    this._showNotification();
+  }
+
+// ...the rest of the class
+```
+
+![Basic desktop notification](/img/blog/add-desktop-notifications/basic-notification.png)
+
+This gets us a basic notification, but without any interaction from the user. To do that, we have to add actions to the options hash that you can watch and respond too. The spec requires that we only can do that if we're triggering the notification via a service worker and responding to events there. Let's create a basic service worker file and add a listener. Create a file called `service-worker.js` and add the following code.
+
+```js
+self.addEventListener('notificationclick', function(event) {
+  const clickedNotification = event.notification;
+  clickedNotification.close();
+
+  // Do something as the result of the notification click
+  const promiseChain = doSomething();
+  event.waitUntil(promiseChain);
+});
+```
+
+In this we're watching for a `notificationclick` event and want to do something with that once we know the user has interacted with it. We'll also want to register our worker in the notification service and have it trigger the notification flow.
+
+```js
+_startServiceWorker(notificationService) {
+  if ('serviceWorker' in navigator) {
+    navigator.serviceWorker.register('service-worker.js')
+    .then(function(registration) {
+      registration.addEventListener('updatefound', function() {
+        // If updatefound is fired, it means that there's
+        // a new service worker being installed.
+        const installingWorker = registration.installing;
+        console.log('A new service worker is being installed:',
+          installingWorker);
+
+        // You can listen for changes to the installing service worker's
+        // state via installingWorker.onstatechange
+      });
+      
+      notificationService._showNotification(registration);
+    })
+    .catch(function(error) {
+      console.log('Service worker registration failed:', error);
+    });
+  } else {
+    console.log('Service workers are not supported.');
+  }
+}
+```
+
+In the above, we check for service worker support and then register our newly created worker file. After it's installled we show the notification in an updated `_showNotification` method. 
+
+```js
+async _showNotification(worker) {
+  const permission = await this._isShowNotificationPermitted();
+  if (permission !== 'denied') {
```

Comment:
It can be confusing to have most of the directions in paragraph form and some as commented out code - it might get missed. Maybe move those commented out lines out and refer to lines A, B, C...